File: app/local_plugins/vendors/llm_volcengine/android/src/main/kotlin/com/aiagent/llm_volcengine/LlmVolcengineService.kt
package com.aiagent.llm_volcengine

import android.util.Log
import com.aiagent.plugin_interface.LlmCallback
import com.aiagent.plugin_interface.NativeLlmService
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import kotlin.coroutines.coroutineContext

/**
 * LlmVolcengineService — 火山引擎 Ark LLM 原生服务实现
 *
 * 走 ark 的 OpenAI 兼容接口：POST <baseUrl>/chat/completions (默认 baseUrl 写死为
 * https://ark.cn-beijing.volces.com/api/v3)；`model` 即接入点 ID（ep-xxx）。
 * 无 configJson / 空 baseUrl 时自动回退默认地址。
 */
class LlmVolcengineService : NativeLlmService {

    companion object {
        private const val TAG = "LlmVolcengineService"
        private const val DEFAULT_BASE_URL =
            "https://ark.cn-beijing.volces.com/api/v3"
    }

    private val client = OkHttpClient()
    @Volatile private var activeCall: Call? = null

    private var apiKey: String = ""
    private var baseUrl: String = DEFAULT_BASE_URL
    private var model: String = ""
    private var temperature: Double = 0.7
    private var maxTokens: Int = 2048
    private var systemPrompt: String? = null
    private var enableThinking: Boolean = false

    override fun initialize(configJson: String) {
        val cfg = JSONObject(configJson)
        apiKey = cfg.optString("apiKey", "")
        baseUrl = normalizeBaseUrl(cfg.optString("baseUrl", ""))
        model = cfg.optString("model", "")
        temperature = cfg.optDouble("temperature", 0.7)
        maxTokens = cfg.optInt("maxTokens", 2048)
        systemPrompt = cfg.optString("systemPrompt", "").ifBlank { null }
        enableThinking = cfg.optBoolean("enableThinking", false)
        Log.d(TAG, "initialize: baseUrl=$baseUrl model=$model thinking=$enableThinking")
    }

    override suspend fun chat(
        requestId: String,
        messages: List<Map<String, Any>>,
        tools: List<Map<String, Any>>,
        callback: LlmCallback,
    ): String {
        if (apiKey.isBlank() || model.isBlank()) {
            val errMsg = "LLM config incomplete: apiKey=${apiKey.isNotBlank()} model=${model.isNotBlank()}"
            Log.e(TAG, errMsg)
            callback.onError("config_error", errMsg)
            return ""
        }

        val messagesArray = JSONArray()
        if (!systemPrompt.isNullOrBlank()) {
            messagesArray.put(JSONObject().apply {
                put("role", "system")
                put("content", systemPrompt)
            })
        }
        for (msg in messages) {
            messagesArray.put(JSONObject(msg))
        }

        val body = JSONObject().apply {
            put("model", model)
            put("stream", true)
            put("messages", messagesArray)
            if (temperature != 0.7) put("temperature", temperature)
            if (maxTokens != 2048) put("max_tokens", maxTokens)
            if (tools.isNotEmpty()) {
                put("tools", JSONArray(tools.map { JSONObject(it) }))
            }
            // 火山方舟 thinking 模型：默认关闭推理，避免首字延迟
            put("thinking", JSONObject().apply {
                put("type", if (enableThinking) "enabled" else "disabled")
            })
        }

        val url = "$baseUrl/chat/completions"

        val request = Request.Builder()
            .url(url)
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .header("Authorization", "Bearer $apiKey")
            .build()

        Log.d(TAG, "POST $url model=$model messages=${messages.size}")

        val fullText = StringBuilder()
        var isFirstToken = true

        return try {
            val response = suspendableEnqueue(request) ?: return ""

            if (!response.isSuccessful) {
                val respBody = response.body?.string()?.take(300) ?: ""
                val errMsg = "HTTP ${response.code}: $respBody"
                Log.e(TAG, "LLM request failed: $errMsg")
                callback.onError("http_${response.code}", errMsg)
                return ""
            }

            val source = response.body?.source() ?: return ""
            // 跨 chunk 跟踪每个 tool_call 是否已经派发过 Start —— OpenAI 流式
            // 协议里同一 tool_call 通过 `index` 复用，第一帧带 id+name，后续
            // 帧只追加 arguments delta。
            val toolStarted = mutableMapOf<Int, Boolean>()

            while (!source.exhausted()) {
                if (!coroutineContext.isActive) break

                val line = source.readUtf8Line() ?: break
                if (!line.startsWith("data: ")) continue
                val data = line.removePrefix("data: ").trim()
                if (data == "[DONE]") break

                val chunk = parseDelta(data) ?: continue

                // 文本
                if (chunk.content.isNotEmpty()) {
                    fullText.append(chunk.content)
                    if (isFirstToken) {
                        isFirstToken = false
                        callback.onFirstToken(chunk.content)
                    } else {
                        callback.onTextDelta(chunk.content)
                    }
                }

                // 推理（doubao thinking 模型走 reasoning_content）
                if (chunk.thinking.isNotEmpty()) {
                    callback.onThinkingDelta(chunk.thinking)
                }

                // tool_calls：首帧带 id+name → onToolCallStart；
                // 后续帧只有 arguments delta → onToolCallArguments
                for (tc in chunk.toolCalls) {
                    val started = toolStarted[tc.index] == true
                    if (!started && !tc.id.isNullOrBlank() && !tc.name.isNullOrBlank()) {
                        callback.onToolCallStart(tc.id, tc.name)
                        toolStarted[tc.index] = true
                    }
                    if (tc.argsDelta.isNotEmpty()) {
                        callback.onToolCallArguments(tc.argsDelta)
                    }
                }
            }

            Log.d(TAG, "stream done: textLen=${fullText.length} toolCalls=${toolStarted.size}")
            callback.onDone(fullText.toString())
            fullText.toString()

        } catch (e: CancellationException) {
            activeCall?.cancel()
            ""
        } catch (e: IOException) {
            Log.e(TAG, "IO error: ${e.message}")
            callback.onError("io_error", e.message ?: "Unknown IO error")
            ""
        }
    }

    override fun cancel() {
        activeCall?.cancel()
    }

    private fun normalizeBaseUrl(raw: String): String {
        var u = raw.trim()
        if (u.isEmpty()) return DEFAULT_BASE_URL
        if (u.endsWith("/")) u = u.trimEnd('/')
        if (u.endsWith("/chat/completions")) {
            u = u.removeSuffix("/chat/completions")
        }
        return u
    }

    private suspend fun suspendableEnqueue(request: Request): Response? {
        return suspendCancellableCoroutine { cont ->
            val call = client.newCall(request)
            activeCall = call
            cont.invokeOnCancellation { call.cancel() }
            call.enqueue(object : Callback {
                override fun onResponse(call: Call, response: Response) {
                    if (cont.isActive) cont.resumeWith(Result.success(response))
                }
                override fun onFailure(call: Call, e: IOException) {
                    if (cont.isActive) cont.resumeWith(Result.failure(e))
                }
            })
        }
    }

    /**
     * 解析单帧 SSE delta。OpenAI 兼容格式：
     * - delta.content：文本增量
     * - delta.reasoning_content：思维链增量（doubao thinking 模型）
     * - delta.tool_calls[]：工具调用片段；首帧 `{index, id, function:{name, arguments:""}}`，
     *   后续帧 `{index, function:{arguments:"...delta..."}}`
     */
    private fun parseDelta(json: String): DeltaChunk? = try {
        val choice = JSONObject(json).getJSONArray("choices").getJSONObject(0)
        val delta = choice.optJSONObject("delta") ?: JSONObject()

        val content = delta.optString("content", "")
        val thinking = delta.optString("reasoning_content", "")

        val toolCalls = mutableListOf<ToolCallChunk>()
        delta.optJSONArray("tool_calls")?.let { arr ->
            for (i in 0 until arr.length()) {
                val tc = arr.optJSONObject(i) ?: continue
                val index = tc.optInt("index", i)
                val id = tc.optString("id", "").ifBlank { null }
                val func = tc.optJSONObject("function")
                val name = func?.optString("name", "")?.ifBlank { null }
                val args = func?.optString("arguments", "") ?: ""
                toolCalls.add(ToolCallChunk(index, id, name, args))
            }
        }

        DeltaChunk(content, thinking, toolCalls)
    } catch (_: Exception) {
        null
    }

    private data class DeltaChunk(
        val content: String,
        val thinking: String,
        val toolCalls: List<ToolCallChunk>,
    )

    private data class ToolCallChunk(
        val index: Int,
        val id: String?,
        val name: String?,
        val argsDelta: String,
    )
}

File: apps/client/local_plugins/vendors/llm_volcengine/ios/Classes/LlmVolcengineService.swift
import Foundation
import ai_plugin_interface
import os.log

/// Volcengine Ark streaming chat completion (HTTP + SSE).
///
/// Mirrors the Android `LlmVolcengineService`. Differences from the
/// OpenAI-compatible base:
///   - Default `baseUrl` falls back to `https://ark.cn-beijing.volces.com/api/v3`
///     when the user leaves it blank.
///   - Adds a `thinking` switch (`{type: "enabled"|"disabled"}`) — defaults to
///     disabled so doubao thinking models do not stall on first token.
///   - Parses `delta.reasoning_content` and forwards it as `onThinkingDelta`.
///   - Tool-call deltas are emitted incrementally (`onToolCallStart` at the
///     first frame carrying both id and name, then `onToolCallArguments` for
///     each subsequent arguments chunk) — also matching Kotlin parity.
public final class LlmVolcengineService: NativeLlmService {

    private static let log = OSLog(subsystem: "com.aiagent.llm_volcengine", category: "Service")
    private static let defaultBaseUrl = "https://ark.cn-beijing.volces.com/api/v3"

    private let session: URLSession
    private let stateLock = NSLock()
    private var activeTask: URLSessionDataTask?

    private var apiKey: String = ""
    private var baseUrl: String = defaultBaseUrl
    private var model: String = ""
    private var temperature: Double = 0.7
    private var maxTokens: Int = 2048
    private var systemPrompt: String?
    private var enableThinking: Bool = false

    public init() {
        let cfg = URLSessionConfiguration.ephemeral
        cfg.timeoutIntervalForRequest = 120
        cfg.timeoutIntervalForResource = 600
        cfg.waitsForConnectivity = false
        self.session = URLSession(configuration: cfg)
    }

    public func initialize(configJson: String) {
        guard let data = configJson.data(using: .utf8),
              let cfg = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            os_log("initialize: invalid config json", log: Self.log, type: .error)
            return
        }
        apiKey = (cfg["apiKey"] as? String) ?? ""
        baseUrl = normalizeBaseUrl((cfg["baseUrl"] as? String) ?? "")
        model = (cfg["model"] as? String) ?? ""
        if let t = cfg["temperature"] as? Double { temperature = t }
        else if let t = cfg["temperature"] as? Int { temperature = Double(t) }
        if let m = cfg["maxTokens"] as? Int { maxTokens = m }
        let sp = (cfg["systemPrompt"] as? String) ?? ""
        systemPrompt = sp.isEmpty ? nil : sp
        enableThinking = (cfg["enableThinking"] as? Bool) ?? false
        os_log("initialize: baseUrl=%{public}@ model=%{public}@ thinking=%{public}@",
               log: Self.log, type: .debug, baseUrl, model, String(describing: enableThinking))
    }

    public func chat(
        requestId: String,
        messages: [[String: Any]],
        tools: [[String: Any]],
        callback: LlmCallback
    ) async throws -> String {
        guard !apiKey.isEmpty, !model.isEmpty else {
            let msg = "LLM config incomplete: apiKey=\(!apiKey.isEmpty) model=\(!model.isEmpty)"
            os_log("%{public}@", log: Self.log, type: .error, msg)
            callback.onError(code: "config_error", message: msg)
            return ""
        }

        var finalMessages: [[String: Any]] = []
        if let sp = systemPrompt, !sp.isEmpty {
            finalMessages.append(["role": "system", "content": sp])
        }
        finalMessages.append(contentsOf: messages)

        var body: [String: Any] = [
            "model": model,
            "stream": true,
            "messages": finalMessages,
            "thinking": ["type": enableThinking ? "enabled" : "disabled"],
        ]
        if abs(temperature - 0.7) > .ulpOfOne { body["temperature"] = temperature }
        if maxTokens != 2048 { body["max_tokens"] = maxTokens }
        if !tools.isEmpty { body["tools"] = tools }

        let url = "\(baseUrl)/chat/completions"
        guard let endpoint = URL(string: url) else {
            callback.onError(code: "config_error", message: "invalid baseUrl: \(url)")
            return ""
        }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONSerialization.data(withJSONObject: body, options: [])

        os_log("POST %{public}@ model=%{public}@ messages=%d tools=%d",
               log: Self.log, type: .debug, url, model, messages.count, tools.count)

        let stream = SseHttpStream(session: session, request: request)
        defer {
            stateLock.lock()
            if activeTask === stream.task { activeTask = nil }
            stateLock.unlock()
        }

        var fullText = ""
        var firstToken = true
        // Track which tool-call indices already emitted `onToolCallStart`.
        // Ark/OpenAI streams reuse a `tool_calls[].index`; the first frame
        // carries `id` + `function.name`, subsequent frames carry only
        // `function.arguments` deltas.
        var toolStarted: [Int: Bool] = [:]

        do {
            try await stream.start()
        } catch {
            callback.onError(code: "io_error", message: error.localizedDescription)
            return ""
        }
        stateLock.lock(); activeTask = stream.task; stateLock.unlock()

        if let http = stream.response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            let bodyPreview = await stream.drainBodyPreview(limit: 300)
            let msg = "HTTP \(http.statusCode): \(bodyPreview)"
            os_log("LLM request failed: %{public}@", log: Self.log, type: .error, msg)
            callback.onError(code: "http_\(http.statusCode)", message: msg)
            return ""
        }

        do {
            for try await line in stream.lines() {
                try Task.checkCancellation()
                guard line.hasPrefix("data: ") else { continue }
                let data = line.dropFirst(6).trimmingCharacters(in: .whitespacesAndNewlines)
                if data == "[DONE]" { break }

                guard let parsed = parseDelta(data) else { continue }

                if let content = parsed.contentDelta, !content.isEmpty {
                    fullText.append(content)
                    if firstToken {
                        firstToken = false
                        callback.onFirstToken(textDelta: content)
                    } else {
                        callback.onTextDelta(content)
                    }
                }

                if let thinking = parsed.thinkingDelta, !thinking.isEmpty {
                    callback.onThinkingDelta(thinking)
                }

                for tc in parsed.toolCallDeltas {
                    let already = toolStarted[tc.index] == true
                    if !already, let id = tc.id, !id.isEmpty, let name = tc.name, !name.isEmpty {
                        callback.onToolCallStart(id: id, name: name)
                        toolStarted[tc.index] = true
                    }
                    if let args = tc.argumentsDelta, !args.isEmpty {
                        callback.onToolCallArguments(args)
                    }
                }
            }
        } catch is CancellationError {
            stream.cancel()
            return ""
        } catch {
            os_log("IO error: %{public}@", log: Self.log, type: .error, error.localizedDescription)
            callback.onError(code: "io_error", message: error.localizedDescription)
            return ""
        }

        os_log("stream done: textLen=%d toolCalls=%d",
               log: Self.log, type: .debug, fullText.count, toolStarted.count)
        callback.onDone(fullText: fullText)
        return fullText
    }

    public func cancel() {
        stateLock.lock()
        let task = activeTask
        activeTask = nil
        stateLock.unlock()
        task?.cancel()
    }

    // ── Parsing helpers ─────────────────────────────────────────────

    private struct DeltaParsed {
        let contentDelta: String?
        let thinkingDelta: String?
        let toolCallDeltas: [ToolCallDelta]
    }

    private struct ToolCallDelta {
        let index: Int
        let id: String?
        let name: String?
        let argumentsDelta: String?
    }

    private func parseDelta(_ jsonString: String) -> DeltaParsed? {
        guard let data = jsonString.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = root["choices"] as? [[String: Any]],
              let first = choices.first else {
            return nil
        }
        let delta = (first["delta"] as? [String: Any]) ?? [:]
        let content: String? = {
            let s = (delta["content"] as? String) ?? ""
            return s.isEmpty ? nil : s
        }()
        let thinking: String? = {
            let s = (delta["reasoning_content"] as? String) ?? ""
            return s.isEmpty ? nil : s
        }()

        var toolDeltas: [ToolCallDelta] = []
        if let tcs = delta["tool_calls"] as? [[String: Any]] {
            for (i, tc) in tcs.enumerated() {
                let fn = tc["function"] as? [String: Any]
                let id = (tc["id"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                let name = (fn?["name"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                let args = (fn?["arguments"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                toolDeltas.append(ToolCallDelta(
                    index: (tc["index"] as? Int) ?? i,
                    id: id,
                    name: name,
                    argumentsDelta: args
                ))
            }
        }
        return DeltaParsed(contentDelta: content, thinkingDelta: thinking, toolCallDeltas: toolDeltas)
    }

    private func normalizeBaseUrl(_ raw: String) -> String {
        var u = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if u.isEmpty { return Self.defaultBaseUrl }
        while u.hasSuffix("/") { u.removeLast() }
        if u.hasSuffix("/chat/completions") {
            u = String(u.dropLast("/chat/completions".count))
            while u.hasSuffix("/") { u.removeLast() }
        }
        return u
    }
}
